Move try_cast() from mssql to base

NickCrews opened the pull request:
Other backends besides mssql support an operation like try_cast. Move it
to a common place so they can all implement it.

Fixes: #9752
Closes: #9752

<!-- Provide a general summary of your proposed changes in the Title field above -->

### Description
<!-- Describe your changes in detail -->

### Checklist
<!-- go over following points. check them with an `x` if they do apply, (they turn into clickable checkboxes once the PR is submitted, so no need to do everything at once)

-->

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [ ] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [x] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

NickCrews committed:
Move try_cast() from mssql to base

Other backends besides mssql support an operation like try_cast. Move it
to a common place so they can all implement it.

Fixes: #9752
Closes: #9752

NickCrews commented:
I only added a deprecation warning if you use `sqlalchemy.dialects.mssql.base.try_cast`. If you use `qlalchemy.dialects.mssql.base.TryCast`, then no deprecation warning. This was partially laziness, I didn't see how to do this easily, and it was partially I don't think most people are using the factory function.

Should I remove the entry from `doc/build/dialects/mssql.rst`? Move it somewhere else?

NickCrews commented on lib/sqlalchemy/dialects/mssql/base.py:
```diff
@@ -1604,40 +1605,14 @@ class SQL_VARIANT(sqltypes.TypeEngine):
+    util.warn_deprecated(
+        "`sqlalchemy.dialects.mssql.base.try_cast` is deprecated. "
+        "Use directly from sqlalchemy instead, i.e. `sa.try_cast(...)`",
+        "2.1",
```
Is this the right version? Should we say when this will be removed?

NickCrews replied:
done

NickCrews commented on lib/sqlalchemy/sql/elements.py:
```diff
@@ -3412,6 +3414,23 @@ def wrapped_column_expression(self):
+    """
+
+    __visit_name__ = "try_cast"
+    inherit_cache = True
```
I just blindly copied these attributes, I don't understand their significance. And I removed `stringify_dialect = "mssql"`, which seems like that probably needs to go somewhere else instead...

NickCrews replied:
That would implement it for ALL backends, even ones that don't support TRY_CAST(), right? and that is fine, the inevitable crash will happen in the right place?

zzzeek replied:
no, StrSQLCompiler is a special compiler that only comes up when you do `str(stmt)`, so that people can see a stringify.  it's not the "main"  compiler that's used for execute().

NickCrews replied:
OK, I think that makes sense. Moved in newest version.

NickCrews replied:
@zzzeek would still love confirmation that I was correct with removing the `stringify_dialect = "mssql"`

zzzeek reviewed (changes requested):
OK so here are the changes:

1. this is not "added new", this is "generalized the MSSQL try_cast function to any dialect, where it can be taken advantage of by third party dialects that also support this function"
2. just move it from mssql, import the "sqlalchemy.sql.elements.try_cast" function into `mssql/base.py` and `mssql/__init__.py` so it still imports the old way.   nothing is "deprecated"
3. add a stringify method (the old MSSQL one) to sqlalchemy/sql/compiler.py -> StrSQLCompiler
4. add a test for this to test/sql/test_compiler.py -> StringifySpecialTest

NickCrews committed:
fixup: try_cast v2

- update changelog entry
- don't deprecate accessing try_cast from mssql. Revert my test change
  to be consistent.
- re-export try_cast from the root place in mssql/__init__.py
- Move visit_try_cast() to base compiler

Still not sure exactly about where the test should live,
since it only should work for mssql, so I'm not sure how to
add it to the generic tests but
specify it to run only for mssql.

NickCrews commented on lib/sqlalchemy/sql/_elements_constructors.py:
```diff
@@ -902,6 +906,30 @@ def cast(
+        SELECT TRY_CAST (product_table.unit_price AS NUMERIC(10, 4))
+        FROM product_table
+
+    .. versionadded:: 2.1.0
```
Should this mention mssql?

is this versionadded correct?

NickCrews replied:
> Should this mention mssql?

Done

I went with 2.0.14 but we can change

NickCrews committed:
fixup: try_cast: move test

NickCrews committed:
fixup: try_cast: make re-exports explicit

CaselIT commented on doc/build/changelog/unreleased_20/9752.rst:
```diff
@@ -0,0 +1,14 @@
+    :tickets: 9752
+
+
+    Generalized the MSSQL :func:`.try_cast` function to any dialect.
```
I think it would make sense to mention that even after this change of the internal dialect only MSSQL supports it

NickCrews replied:
done

CaselIT commented on lib/sqlalchemy/sql/compiler.py:
```diff
@@ -2777,6 +2777,12 @@ def visit_cast(self, cast, **kwargs):
         )
 
+    def visit_try_cast(self, cast, **kwargs):
+        return "TRY_CAST(%s AS %s)" % (
```
Do we want to support it for all dialect, or should we keep that only mssql can compile it?

Or maybe just implement it on the strsqlcompiler?

NickCrews replied:
(caveat that I don't understand the sqlalchemy source) Yeah I'm worried here that if this is in the base compiler then sqlite will happily generate some bad SQL, and the error will only happen at run time. I would think it would be better for the error to happen at compile time.

zzzeek replied:
it should be only in strsqlcompiler.  for sqlcompiler it's not implemented

NickCrews replied:
Ahhh I see, I thought you had a typo when you suggest StrSqlCompiler before, I somehow didn't see that class. Moving it now...

CaselIT reviewed (changes requested):
looks mostly ok!

NickCrews committed:
fixup: try_cast: v5

- Tweak docs to be explicit that only MSSQL
  is currently supported.
- Update try_cast() docstring to better follow
the conventions of cast()
(and just be better in general)
- export try_cast from sql/__init__.py

NickCrews commented on lib/sqlalchemy/__init__.py:
```diff
@@ -99,6 +99,8 @@
 from .sql.expression import Cast as Cast
 from .sql.expression import cast as cast
+from .sql.expression import TryCast as TryCast
+from .sql.expression import try_cast as try_cast
```
done in v5 :)

zzzeek commented on lib/sqlalchemy/sql/compiler.py:
```diff
@@ -2777,6 +2777,12 @@ def visit_cast(self, cast, **kwargs):
             cast.typeclause._compiler_dispatch(self, **kwargs),
         )
 
+    def visit_try_cast(self, cast, **kwargs):
```
move this method to the "class StrSQLCompiler" class

NickCrews committed:
fixup: try_cast: move visit_try_cast() to StrSQLCompiler

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 16f1ab1c306c5d46da9a7c816f700237b0a8696d of this pull request into gerrit so we can run tests and reviews and stuff

CaselIT commented on lib/sqlalchemy/sql/_elements_constructors.py:
```diff
@@ -907,24 +908,34 @@ def cast(
+    construct, and works in the same way, except in regards to error
+    handling. If the function encounters an un-castable value
+    (for instance when trying to convert the string "hi" to an INT)
+    then normal :class:`.Cast` raises an error. :class:`.TryCast`
```
```suggestion
    then normal :class:`.Cast` raises an error, :class:`.TryCast`
```

NickCrews replied:
done.

CaselIT reviewed (commented):
thanks for the update!

Also feel free to add "Pull request courtesy <your name>" in the changelog doc

NickCrews committed:
fixup: give me credit in changelog

NickCrews committed:
fixup: docstring tweak

NickCrews committed:
fixup: format with pre-commit

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 41bf05ce6b7d58bebefaa9320c7dec31c70f6868 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 41bf05ce6b7d58bebefaa9320c7dec31c70f6868 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset 41bf05ce6b7d58bebefaa9320c7dec31c70f6868 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4605

NickCrews committed:
fixup: make tests work

NickCrews committed:
fixup: mypy

NickCrews commented:
I hadn't bothered to try to figure out mypy and pytest locally, but I just did and now
- `mypy lib` passes
- `pytest test/sql/test_compiler.py -k test_try_cast` passes. IDK about pytest in general
curios if CI finds other errors, if so I will dig deeper. Thanks @CaselIT for babysitting this!

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 4c0730aa62ed396c106e77509e6f49b10f4e4448 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset 4c0730aa62ed396c106e77509e6f49b10f4e4448 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4605

NickCrews committed:
fixup: indentation

NickCrews commented:
OK I expect all those checks will pass. Want me to squash the commits or anything?

CaselIT commented:
> Want me to squash the commits or anything?

no problem, girrit will do that for us

NickCrews committed:
fixup: run black correctly

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision e6cdaabe114efb6b8c770b9b89475c663cbd2971 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset e6cdaabe114efb6b8c770b9b89475c663cbd2971 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4605

NickCrews committed:
fixup: try_cast: restore mssql implementation

From comments at
https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4605

sqla-tester commented on test/dialect/mssql/test_compiler.py:
```diff
@@ -1474,9 +1474,7 @@ def test_index_include_where(self):
         )
 
     def test_try_cast(self):
```
**Nick Crews** (NickCrews) wrote:

Ack

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4605

sqla-tester reviewed (commented):
**Nick Crews** (NickCrews) wrote:

Thanks Federico, you're totally right. I restored the implementation and test in the newest version.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4605

* lib/sqlalchemy/dialects/mssql/base.py (line 2181): Ack

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision e2be28a4847937203ab6adc4b83a2799164dfcae of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision e2be28a4847937203ab6adc4b83a2799164dfcae of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision e2be28a4847937203ab6adc4b83a2799164dfcae of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset e2be28a4847937203ab6adc4b83a2799164dfcae added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4605

CaselIT commented:
3rd time the charm, I guess. @zzzeek not sure if re-trying is the appropriate thing to do here

zzzeek commented:
> 3rd time the charm, I guess. @zzzeek not sure if re-trying is the appropriate thing to do here

wow i have no idea why the other two didnt work

zzzeek commented:
> > 3rd time the charm, I guess. @zzzeek not sure if re-trying is the appropriate thing to do here
> 
> wow i have no idea why the other two didnt work

90% sure it's more stale github rsa keys in authorized_keys

NickCrews commented:
This has a +1 on Gerrit and all CI passes, friendly bump for any potential reviewers

sqla-tester commented on doc/build/changelog/unreleased_20/9752.rst:
```diff
@@ -0,0 +1,18 @@
+    extension dialects can use it (MSSQL is the only builtin dialect
+    that currently supports it).
+
+    This implements a cast where
```
**mike bayer** (zzzeek) wrote:

I can fix his later, but slight nit: word wrap the paragraph evenly to 79 chars, or I'll just do it later..

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4605

sqla-tester commented on lib/sqlalchemy/dialects/mssql/__init__.py:
```diff
@@ -39,11 +38,11 @@
 from .base import VARBINARY
 from .base import VARCHAR
 from .base import XML
+from ...sql._elements_constructors import try_cast
```
**mike bayer** (zzzeek) wrote:

this can be imported as 

    from ...sql import try_cast as try_cast

that is let's not reach into a private module...

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4605

sqla-tester commented on lib/sqlalchemy/dialects/mssql/base.py:
```diff
@@ -939,8 +939,10 @@ class MyClass(Base):
 from ...sql import roles
 from ...sql import sqltypes
 from ...sql import util as sql_util
+from ...sql._elements_constructors import try_cast as try_cast  # noqa: F401
```
**mike bayer** (zzzeek) wrote:

this can be imported as 

    from ...sql import try_cast as try_cast

that is let's not reach into a private module...

also we are exporting the name (e.g. from X import Y as Z), do we need the "F401" still?  if we do, we do...just curious

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4605

sqla-tester commented on lib/sqlalchemy/sql/_elements_constructors.py:
```diff
@@ -902,6 +907,43 @@ def cast(
+        SELECT TRY_CAST (product_table.unit_price AS NUMERIC(10, 4))
+        FROM product_table
+
+    .. versionadded:: 2.0.14
```
**mike bayer** (zzzeek) wrote:

.. versionadded:: 2.0.14 - the :func:`.try_cast` construct has been generalized from the SQL Server dialect into a general use construct that may be supported by additional dialects.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4605

sqla-tester reviewed (commented):
**mike bayer** (zzzeek) wrote:

a few little changes if you want to do one more go around

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4605

NickCrews committed:
fixup: linewrap changelog blurb better

NickCrews committed:
fixup: Add more explanation to versionadded::

NickCrews committed:
fixup: Import from non-private

sqla-tester commented on doc/build/changelog/unreleased_20/9752.rst:
```diff
@@ -0,0 +1,18 @@
+    dialects can use it (MSSQL is the only builtin dialect that currently
+    supports it).
+
+    This implements a cast where un-castable values are returned as NULL,
```
**Nick Crews** (NickCrews) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4605

sqla-tester commented on lib/sqlalchemy/dialects/mssql/__init__.py:
```diff
@@ -39,11 +38,11 @@
 from .base import VARBINARY
 from .base import VARCHAR
 from .base import XML
+from ...sql import try_cast
```
**Nick Crews** (NickCrews) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4605

sqla-tester commented on lib/sqlalchemy/dialects/mssql/base.py:
```diff
@@ -938,9 +938,11 @@ class MyClass(Base):
 from ...sql import roles
 from ...sql import sqltypes
+from ...sql import try_cast as try_cast  # noqa: F401
 from ...sql import util as sql_util
```
**Nick Crews** (NickCrews) wrote:

We do need the noqa still (per `pre-commit run --files lib/sqlalchemy/dialects/mssql/base.py`, not sure if this is the canonical way to lint).

But I think you might be right that I think there is some config in flake8 that we could turn on so this wasn't needed.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4605

sqla-tester commented on lib/sqlalchemy/sql/_elements_constructors.py:
```diff
@@ -902,6 +907,44 @@ def cast(
+        SELECT TRY_CAST (product_table.unit_price AS NUMERIC(10, 4))
+        FROM product_table
+
+    .. versionadded:: 2.0.14  The :func:.try_cast construct has been
```
**Nick Crews** (NickCrews) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4605

sqla-tester reviewed (commented):
**Nick Crews** (NickCrews) wrote:

All fixed, thanks Mike!

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4605

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 2e81b8d2c9314ca2cc5807fc807c35fefbe6ad2c of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset 2e81b8d2c9314ca2cc5807fc807c35fefbe6ad2c added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4605

sqla-tester reviewed (commented):
**mike bayer** (zzzeek) wrote:

made another pass over docs etc.   There's a lot of explanation and things to go with this one simple move of a function.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4605

sqla-tester commented:
**mike bayer** (zzzeek) wrote:

OK, we're done.  thank you very much!

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4605

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4605 has been **merged**. Congratulations! :)

NickCrews commented:
Thank you @zzzeek ! Those final tweaks look good.
